{% extends "base.html" %}

{% block content %}

    <div class="container">

        <div class="container-fluid text-center page-header">
            <h1>{{ title }}</h1>
            <p class="lead">HmtPhenome aims at providing comprehensive information about variants, genes, phenotypes and diseases involved in mitochondrial functionality.</p>
        </div>

        <div class="container-fluid">
            <div class="row">

                <div class="col-sm-12">
                    <h3>Overview</h3>
                    <p>HmtPhenome is a project framed within the ELIXIR Implementation Study <em>"ELIXIR annotation and curation of human genomic variations"</em>, aimed to support the interpretation of high-throughput human genome sequencing results by integrating different layers of data.</p>
                    <p>Well-assessed services that provide data about human mitochondrial genomes and variants already exist, such as <a href="https://www.hmtdb.uniba.it" title="HmtDB" target="_blank">HmtDB</a> and <a href="https://www.hmtvar.uniba.it" title="HmtVar" target="_blank">HmtVar</a> respectively, but the missing link is a resource specifically focused on collecting phenotypic information from patients affected by any disease in which the mitochondrial DNA has a primary or secondary role.</p>
                    <p>HmtPhenome aims at filling in this gap, integrating data about variants, genes, phenotypes and diseases in which the mitochondrion is involved, and providing and interconnected network view of these entities that would be of great help for researchers and clinicians interested in deepening their interpretation of human genomics by integrating different layers of data.</p>
                </div>

                <div class="col-sm-12">
                    <h3>Querying HmtPhenome</h3>
                    <p>HmtPhenome gathers data from several different resources available online, and these data are then integrated to build a knowledge network of entities, which are interconnected and allow to retrieve a comprehensive scenario of the mitochondrial involvement as related to genes, variants, diseases and phenotypes of interest. Users can query the system starting from any one of these entry points, and retrieve information pertaining to the other 3 related entities.</p> 
                    <p>External resources from which data are collected include <a href="https://www.ensembl.org/index.html" title="Ensembl" target="_blank">Ensembl</a>, <a href="https://hpo.jax.org" title="HPO" target="_blank">HPO</a>, <a href="https://www.ebi.ac.uk/efo/" title="EFO" target="_blank">EFO</a>, <a href="https://www.omim.org" title="OMIM" target="_blank">OMIM</a> and <a href="https://www.orpha.net/consor/cgi-bin/index.php" title="Orphanet" target="_blank">Orphanet</a>.</p>
                    <p>When a user queries HmtPhenome, all required data are retrieved from these online resources and integrated on the fly. The only data coming from a local database are the list of human mtDNA-encoded genes and nuclear-encoded genes with mitochondrial involvements, retrieved from <a href="https://www.broadinstitute.org/files/shared/metabolism/mitocarta/human.mitocarta2.0.html" title="MitoCarta" target="_blank">MitoCarta</a>, and the list of diseases and phenotypes with a mitochondrial component.</p>
                </div>

                <div class="col-sm-12">
                    <h3>Implementation</h3>
                    <p>HmtPhenome retrieves data from several external resources when a query is launched, so it is pivotal that this system be fast and efficient, avoiding long running times. For this reason, HmtPhenome is built upon the <a href="https://github.com/pgjones/quart" title="Quart" target="_blank">Quart</a> Python framework, which is specifically suited to address this issue. All requests issued to external resources are handled separately and in parallel, so the waiting time experienced by the end user is definitely smaller than with classic data retrieval systems.</p>
                </div>

            </div>
        </div>

    </div>

    {{ super() }}
{% endblock %}
